services/paymaster/src/api.rs

```diff
@@ -59,6 +62,57 @@ pub fn validate_transaction(
             }
             Ok(())
         })?;
+
+    // Simulate the transaction to check sponsor SOL spending
+    let simulation_result = rpc
+        .simulate_transaction_with_config(
+            transaction,
+            RpcSimulateTransactionConfig {
+                sig_verify: false,
+                replace_recent_blockhash: true,
+                ..RpcSimulateTransactionConfig::default()
+            },
+        )
+        .map_err(|err| {
+            (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                format!("Failed to simulate transaction: {err}"),
+            )
+        })?;
+
+    if let Some(err) = simulation_result.value.err {
+        return Err((
+            StatusCode::INTERNAL_SERVER_ERROR,
+            format!("Transaction simulation failed: {err:?}"),
+        ));
+    }
+
+    // Check if the sponsor account balance change exceeds the maximum permissible value
+    if let Some(accounts) = simulation_result.value.accounts {
+        if let Some(Some(account)) = accounts.first() {
```

Comment:
what's with the double wrapping?

Reply:
accounts is `Vec<Option<Account>>` (defined by the rpc) and so first returns `Option<Option<Account>>`.